data_steward/cdr_cleaner/cleaning_rules/drop_ppi_duplicate_responses.py

```diff
@@ -0,0 +1,214 @@
+"""
+Removes the duplicate sets of responses to the same questions excluding COPE survey responses.
+
+Original Issues: DC-1051, DC-532
+
+In PPI surveys, the purpose of questionnaire_response_id is to group all responses from the same survey together.
+Some PPI questions allowed participants to provide multiple answers, which be will connected via the same
+questionnaire_response_id. However, a participant may submit the responses multiple times for the same questions,
+therefore creating duplicates. We need to use the combination of person_id, observation_source_concept_id,
+observation_source_value, and questionnaire_response_id to identify multiple sets of responses. We only want to keep
+the most recent set of responses and remove previous sets of responses. When identifying the most recent responses,
+we can't use questionnaire_response_id alone because a larger questionnaire_response_id doesn't mean it's created at
+a later time therefore we need to create ranks (lowest rank = most recent responses) based on observation_date_time.
+However, we also need to add questionnaire_response_id for assigning unique ranks to different sets of responses
+because there are cases where the multiple sets of responses for the same question were submitted at exactly the same
+timestamp but with different answers. In addition, we also need to check whether one of the duplicate responses is
+PMI_Skip because there are cases where the most recent response is a skip and the legitimate response was submitted
+earlier, we want to keep the actual response instead of PMI_Skip regardless of the timestamps of those responses.
+"""
+
+# Python imports
+import logging
+
+# Third party imports
+from jinja2 import Environment
+
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+
+LOGGER = logging.getLogger(__name__)
+
+jinja_env = Environment(
+    # help protect against cross-site scripting vulnerabilities
+    autoescape=True,
+    # block tags on their own lines
+    # will not cause extra white space
+    trim_blocks=True,
+    lstrip_blocks=True,
+    # syntax highlighting should be better
+    # with these comment delimiters
+    comment_start_string='--',
+    comment_end_string=' --')
+
+PIPELINE_DATASET = 'pipeline_tables'
+COPE_CONCEPTS_TABLE = 'cope_concepts'
+
+SELECT_DUPLICATE_TEMPLATE = jinja_env.from_string("""
+CREATE OR REPLACE TABLE
+    `{{project}}.{{sandbox_dataset}}.{{intermediary_table}}` AS 
+SELECT
+  *
+FROM
+  `{{project}}.{{dataset}}.observation`
+WHERE
+  observation_id IN 
+""")
+
+REMOVE_DUPLICATE_TEMPLATE = jinja_env.from_string("""
+DELETE
```

Comment:
Should this be updated to not use a DML statement? Replace with `SELECT` and `WRITE TRUNCATE`?

Reply:
@amcgrenera-vumc Because it is affecting a large amount of data and either of them is performing similarly, so I left it as is. I can update it if it is needed.

Reply:
No if it makes more sense to leave as `DELETE` that's fine!